prog/checksum.go

```diff
@@ -120,38 +121,114 @@ func encodeStruct(arg *Arg, pid int) []byte {
 	return bytes
 }
 
-func calcChecksumIPv4(arg *Arg, pid int) (*Arg, *Arg) {
-	var csumField *Arg
+func findFieldByName(arg *Arg, name string) *Arg {
 	for _, field := range arg.Inner {
-		if _, ok := field.Type.(*sys.CsumType); ok {
-			csumField = field
-			break
+		if field.Type.FieldName() == name {
+			return field
 		}
 	}
+	return nil
+}
+
+func calcChecksumIPv4(arg *Arg, pid int) (*Arg, *Arg) {
+	csumField := findFieldByName(arg, "csum")
 	if csumField == nil {
```

Comment:
I would move the nil check and panic into findFieldByName itself, because it is duplicated multiple times.
It is also possible to type/size checks into findFieldByName. If some of the checks are optional, then you can use 0/nil as 'don't check'.